internal/core/components/watcher.go

```diff
@@ -0,0 +1,192 @@
+				continue
+			}
+
+			wg := sync.WaitGroup{}
+			wg.Add(len(values))
+			for _, val := range values {
+				w.sem <- struct{}{}
+				go w.handleValue(val, &wg, s)
+			}
+			wg.Wait()
+		}
+	}
+}
+
+func (w *Watcher) handleValue(val []string, wg *sync.WaitGroup, s *storer.GenericStore) {
+	defer func() {
+		<-w.sem
+		wg.Done()
+	}()
+
+	log.Infof("Watcher handle %s event: key=%s value=%s", val[0], val[1], val[2])
+	switch val[0] {
+	case utils.EventAdd:
+		objPtr, err := s.StringToObjPtr(val[2], val[1])
+		if err != nil || !entity.ServiceFilter(objPtr) {
+			return
+		}
+		queer := objPtr.(entity.Queer)
+
+		oldObjPtr, ok := s.Store(val[1], objPtr)
+		if !ok {
+			// Obtains a new entity with service information
+			log.Info("Watcher obtains a new entity with service information")
+			query, err := encodeQuery(utils.EventAdd, s.Typ, queer)
+			if err != nil {
+				return
+			}
+			_ = discoverer.GetDiscoverer(queer.GetType()).Query(query)
+		} else if entity.ServiceUpdate(oldObjPtr, objPtr) {
+			// Updates the service information of existing entity
+			log.Info("Watcher updates the service information of existing entity")
+			update, err := encodeUpdate(oldObjPtr.(entity.Queer), queer)
+			if err != nil {
+				return
+			}
+			_ = discoverer.GetDiscoverer(queer.GetType()).Update(update)
+		} else if entity.ServiceReplace(oldObjPtr, objPtr) {
+			// Replaces the service information of existing entity
+			log.Info("Watcher replaces the service information of existing entity")
+			oldQueer := oldObjPtr.(entity.Queer)
+			del, err := encodeQuery(utils.EventDelete, s.Typ, oldQueer)
+			if err != nil {
+				return
+			}
+			add, err := encodeQuery(utils.EventAdd, s.Typ, queer)
+			if err != nil {
+				return
+			}
+
+			_ = discoverer.GetDiscoverer(oldQueer.GetType()).Query(del)
+			_ = discoverer.GetDiscoverer(queer.GetType()).Query(add)
+		}
+	case utils.EventDelete:
+		objPtr, ok := s.Delete(val[1])
+		if ok {
+			// Deletes an existing entity
+			log.Info("Watcher deletes an existing entity")
+			queer := objPtr.(entity.Queer)
+			query, err := encodeQuery(utils.EventDelete, s.Typ, queer)
+			if err != nil {
+				return
+			}
+			_ = discoverer.GetDiscoverer(queer.GetType()).Query(query)
+		}
+	}
+}
+
+func encodeQuery(event, typ string, queer entity.Queer) (*comm.Query, error) {
+	id, service, args := queer.Extract()
+	entityID := discoverer.EncodeEntityID(typ, id)
+
+	headerVals := []string{event, entityID, service}
+	query, err := comm.NewQuery(headerVals, args)
+	if err != nil {
+		log.Error("Watcher encode query message error")
+		return nil, err
+	}
+
+	return &query, nil
+}
+
+func encodeUpdate(oldQueer, newQueer entity.Queer) (*comm.Update, error) {
+	_, service, args := oldQueer.Extract()
+	_, _, newArgs := newQueer.Extract()
+
+	headerVals := []string{utils.EventUpdate, service}
+	update, err := comm.NewUpdate(headerVals, args, newArgs)
+	if err != nil {
+		log.Error("Watcher encode update message error")
```

Comment:
I think it would be better to let the function which consumes the err (don't throw it out) log down the err. Therefore, we won't miss the err log. Also, when you log the err, you should log down its detail.